src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,342 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <sys/wait.h>
+
+#include "bpf-dlopen.h"
+#include "fileio.h"
+#include "sd-daemon.h"
+#include "sd-messages.h"
+#include "bpf-program.h"
+#include "bpfd-manager.h"
+#include "fd-util.h"
+#include "fs-util.h"
+#include "hashmap.h"
+#include "missing_syscall.h"
+#include "mkdir.h"
+#include "process-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "stdio-util.h"
+#include "umask-util.h"
+
+static int get_memlock_by_fd(int fd, char **memlock) {
+        _cleanup_free_ char *fdinfo = NULL;
+        char *p;
+        char path[STRLEN("/proc/self/fdinfo/") + DECIMAL_STR_MAX(int)];
+        int r;
+
+        xsprintf(path, "/proc/self/fdinfo/%i", fd);
+        r = read_full_virtual_file(path, &fdinfo, NULL);
+        if (r < 0)
+                return r;
+
+        p = find_line_startswith(fdinfo, "memlock:");
+        if (!p)
+                return -EOPNOTSUPP;
+
+        p += strspn(p, WHITESPACE);
+        p[strcspn(p, WHITESPACE)] = 0;
+
+        *memlock = strdup(p);
+        if (!*memlock)
+                return -ENOMEM;
+
+        return 0;
+}
+
+static int list_bpf_maps_and_progs(Manager *m) {
+        int i = 0;
+        uint32_t id = 0;
+        struct bpf_map_info map_info = {};
+        uint32_t map_info_len = sizeof(map_info);
+        struct bpf_prog_info prog_info = {};
+        uint32_t prog_info_len = sizeof(prog_info);
+        uint32_t nr_map_ids = 0;
+        struct bpf_func_info func_info = {};
+        uint32_t func_info_rec_size = 0;
+        const struct btf_type *t;
+        struct btf *btf = NULL;
+        char prog_tag[BPF_TAG_SIZE * 2 + 1];
+        _cleanup_hashmap_free_ Hashmap *map_name_by_id = NULL;
+        int r;
+
+        map_name_by_id = hashmap_new(NULL);
+        if (!map_name_by_id)
+                return -ENOMEM;
+        /* Iterate through all BPF maps to get map names, map types, and memlock */
+        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char *memlock = NULL;
+                char *map_name = NULL;
+                uint32_t map_id = id;
+
+                /* Grab the FD using the ID*/
+                fd = sym_bpf_map_get_fd_by_id(id);
+                if (fd < 0)
+                        return log_error_errno(errno, "Can't get fd for id: %u, ignoring", id);
```

Comment:
If we're just ignoring here do we want to return or rather just log and continue iterating?

Reply:
I think we can just return, I put "ignoring" while testing but forgot to remove it.

Reply:
After more discussion, we've decided that just continuing will probably be OK. Will update most of the error handling in the code to reflect this.